File: Java/src/Queue/CircularArrayBaseQueue.java
package Queue;

import java.util.Iterator;

public class CircularArrayBaseQueue<T> implements QueueADT<T> {
    private final T[] array;
    private int front;
    private int end;
    private final int size;

    public CircularArrayBaseQueue(int size) {
        this.size = size + 1;
        this.array = (T[]) new Object[this.size];
        this.front = 0;
        this.end = 0;
    }

    @Override
    public void enQueue(T element) {
        array[end] = element;
        if (++end == size) end = 0;
        if (end == front) throw new RuntimeException("Queue is full!");
    }

    @Override
    public T deQueue() {
        if (isEmpty()) throw new RuntimeException("Queue empty!");
        T enQueueElem = array[front];
        array[front] = null;
        if (++front == size) front = 0;
        return enQueueElem;
    }

    @Override
    public T peek() {
        if (isEmpty()) throw new RuntimeException("Queue empty!");
        return array[front];
    }

    @Override
    public int size() {
        if (front > end) return end + size - front;
        return end - front;
    }

    @Override
    public boolean isEmpty() {
        return front == end;
    }

    @Override
    public Iterator<T> iterator() {
        return null;
    }
}

File: Typescript/src/Queue/CircularArrayBaseQueue.ts
import { QueueADT } from "./QueueADT";

export class ArrayBaseQueue<T> implements QueueADT<T> {
  private readonly array: T[];
  private front: number;
  private end: number;
  private readonly size: number;

  constructor(maxSize: number) {
    this.front = this.end = 0;
    this.size = maxSize + 1;
    this.array = new Array(this.size);
  }

  enQueue(element: T): void {
    this.array[this.end] = element;
    if (++this.end == this.size) this.end = 0;
    if (this.end == this.front) throw new Error("Queue is full");
  }

  deQueue(): T | null {
    if (this.isEmpty()) throw new Error("Queue is empty");
    let deQueueElement: T | null = this.array[this.front];
    deQueueElement = null;
    if (++this.front == this.size) this.front = 0;
    return deQueueElement;
  }
  peek(): T | null {
    if (this.isEmpty()) throw new Error("Queue is empty");
    return this.array[this.front];
  }
  capacity(): number {
    if (this.front > this.end) return this.end + this.size - this.front;
    return this.end - this.front;
  }
  isEmpty(): boolean {
    return this.front == this.end;
  }
}
